### DISCUSSION TITLE
Implementing SQIP (SVG quality image previews) -> pixel data to SVG polygons

### DISCUSSION THREAD

**Discussion by kipcole9** [libvips/libvips#3115] [Help wanted]
I wonder if there is any prior art on using `libvips` to take a raster image and convert it to a (lower quality) SVG-based image?  SQIP is described in [this artcle](https://calendar.perfplanet.com/2017/sqip-vague-vectors-for-performant-previews/) and there is a very interesting `golang` implementation called [primitive](https://github.com/fogleman/primitive) that best illustrates the intent and describes the algorithm.

By definition this process is going to be compute-heavy. Some additional things that may possibly make `libvips` not the right place to start:

1. Calculating the root-mean-square-error between two images (frequently)
2. Compositing shapes (via SVG renders) isn't going to be fast and thats the core of the algorithm. I don't think the `vips_draw_*` functions are going to be a good strategy either (and the shapes are limited)

Thoughts, comments, suggestions?

One alternative approach might be the quantise the image (I think there may be some quantise functions coming in 8.14 that aren't bound to file operations?), find the edges and convert them to SVG but I don't think that will produce as good an image as the process used by `primitive`.

**Comment by jcupitt** [comment]
Hi @kipcole9,

I played with `primitive` a bit when it came out, it's a nice thing.  You might be able to make it quicker by replacing some of the pixel arithmetic with libvips, but I've not tried. 

Perhaps profile it, pick the most obvious hotspot that could be replaced, reimplement it, and benchmark again? It wouldn't take too long, and you'd get some feeling about whether any gains are possible and could be worthwhile.
